### ISSUE TITLE
Rename keywords for clarity?

### ISSUE THREAD

**Issue by alphapapa** [jwiegley/use-package#382]
This is just something to think about.  I know it would be a breaking change.  But I think it would be useful for the sake of clarity,  so worth considering.  :)

When I'm thinking about `use-package`, I can never remember whether `:init` and `:config` run before or after a package loads; I have to refer to the help to refresh my memory.  Just from the term itself, `init` could mean "before the package initializes," or "after the package initializes," or even "to initialize the package" (which is ambiguous).

If they were named differently, their function could be self-evident.  

For example, if `:init` were named `:before`, it would be obvious that its code ran before the package loaded.  (Other possibilities could be "`:prepare`, `:pre`, `:pre-load` (not `preload` which would seem to mean that the package should be "preloaded"), `:prologue`, `:preface` (oops, that one's already taken too...), etc.)

If `:config` were named `:after`, it would be obvious that its code ran after the package loaded.  (Other possibilities: `:post`, `:post-load`, `:epilogue`, `:afterword`, etc.)

I'm aware that there is already an `:after` keyword, but I think that one would be better named `:defer-until`, as in "Defer loading of a package until after any of the named features are loaded" (I copied that from the help ;).

In conclusion, naming things is hard.

**Comment by Silex** [comment]
:+1:  for `:before` and `:after`

As far as I am concerned, as the current `:after` is not documented in the README it's ok to rename it, because not many people are using it.

That said, I'm ok with `:pre-load`, `:post-load` and `:defer-until` too, with `:init`, `:config` and `:after` marked as deprecated for a while.

**Comment by jwiegley** [comment]
This has been discussed in some depth already in #306.

You can always define these keywords for yourself, immediately after loading `use-package`:

``` elisp
(setq use-package-keywords
      (mapcar
       #'(lambda (kw)
           (cl-case kw
             (:init :before)
             (:config :after)
             (otherwise kw)))
       use-package-keywords))

(defalias use-package-normalize/:before 'use-package-normalize/:init)
(defalias use-package-handler/:before 'use-package-handler/:init)

(defalias use-package-normalize/:after 'use-package-normalize/:config)
(defalias use-package-handler/:after 'use-package-handler/:config)
```

The joys of the extensible keyword support in 2.x. :)

The reason that I don't like `:before` and `:after` is because they are misnomers.  A `:before` form is not guaranteed to run "before" the package loads.  If the package was already loaded before the `use-package` declaration, it will run after.

**Comment by alphapapa** [comment]
I looked at #306 and saw this comment by @dabrahams:

> P.S. I'd love it if init would be renamed “immediately” and “config” would be renamed “before-load” or “after-load” as appropriate to its semantics, whatever they may be.  I find it really hard to remember
these things and have to go back to the docs every time I want to add or maintain a new invocation.

Hm, same problem I'm having.  ;)  I doubt we're the only two people who find this confusing, so I think it would be very helpful to change these keywords or add more descriptive aliases for them.

> The reason that I don't like :before and :after is because they are misnomers. A :before form is not guaranteed to run "before" the package loads. If the package was already loaded before the use-package declaration, it will run after.

I understand they're not guaranteed, but I don't understand the objection.  Here's the docstring:

```
:init          Code to run before PACKAGE-NAME has been loaded.
:config        Code to run after PACKAGE-NAME has been loaded.  Note that if
               loading is deferred for any reason, this code does not execute
               until the lazy load has occurred.
```

How is that better than, e.g.:

```
:before          Code to run before PACKAGE-NAME has been loaded.
:after        Code to run after PACKAGE-NAME has been loaded.  Note that if
               loading is deferred for any reason, this code does not execute
               until the lazy load has occurred.
```

:)
